File: android/app/src/main/java/com/metalens/app/upload/AwsV4Signer.kt
package com.metalens.app.upload

import java.security.MessageDigest
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * Minimal AWS Signature Version 4 signer, scoped to exactly what the S3 uploader needs:
 * one request, a fully buffered payload, and no query string.
 *
 * Hand-rolled rather than pulled from the AWS SDK: the SDK drags in a large transitive tree
 * (smithy-kotlin, kotlinx-serialization, its own HTTP engine) for a single PUT, while the rest
 * of this app already speaks raw OkHttp. See AwsV4SignerTest for the AWS-published test vector.
 */
object AwsV4Signer {
    private const val ALGORITHM = "AWS4-HMAC-SHA256"
    private const val TERMINATOR = "aws4_request"

    /** RFC 3986 unreserved characters — everything else in a path segment is percent-encoded. */
    private const val UNRESERVED =
        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~"

    /**
     * Signs the request and returns the headers to add to it.
     *
     * [headers] are the extra headers to sign (content-type, x-amz-meta-*, x-amz-security-token).
     * The returned map also contains `x-amz-date`, `x-amz-content-sha256` and `Authorization`.
     * `Host` is signed but not returned, because OkHttp derives it from the URL itself.
     */
    fun sign(
        method: String,
        host: String,
        /** Raw, *unencoded* object key with no leading slash. */
        key: String,
        headers: Map<String, String>,
        payloadSha256: String,
        accessKeyId: String,
        secretAccessKey: String,
        region: String,
        service: String = "s3",
        timestamp: Date,
    ): Map<String, String> {
        val amzDate = amzDateFormat().format(timestamp)
        val dateStamp = dateStampFormat().format(timestamp)

        val signed =
            headers +
                mapOf(
                    "host" to host,
                    "x-amz-content-sha256" to payloadSha256,
                    "x-amz-date" to amzDate,
                )

        val canonical = canonicalRequest(method, key, signed, payloadSha256)
        val scope = "$dateStamp/$region/$service/$TERMINATOR"
        val stringToSign = stringToSign(amzDate, scope, canonical)

        val signingKey = signingKey(secretAccessKey, dateStamp, region, service)
        val signature = hex(hmacSha256(signingKey, stringToSign))

        val authorization =
            "$ALGORITHM Credential=$accessKeyId/$scope, " +
                "SignedHeaders=${signedHeaderNames(signed)}, " +
                "Signature=$signature"

        // Host is excluded: OkHttp sets it from the URL, and setting it twice would break the match.
        return signed.filterKeys { !it.equals("host", ignoreCase = true) } +
            mapOf("Authorization" to authorization)
    }

    internal fun canonicalRequest(
        method: String,
        key: String,
        headers: Map<String, String>,
        payloadSha256: String,
    ): String {
        val canonicalHeaders =
            headers
                .map { (name, value) -> name.lowercase(Locale.US) to normalizeHeaderValue(value) }
                .sortedBy { it.first }
                .joinToString("") { (name, value) -> "$name:$value\n" }

        return listOf(
            method,
            "/" + encodeKey(key),
            // No query string on a plain PUT Object.
            "",
            canonicalHeaders,
            signedHeaderNames(headers),
            payloadSha256,
        ).joinToString("\n")
    }

    internal fun stringToSign(
        amzDate: String,
        scope: String,
        canonicalRequest: String,
    ): String =
        listOf(
            ALGORITHM,
            amzDate,
            scope,
            sha256Hex(canonicalRequest.toByteArray(Charsets.UTF_8)),
        ).joinToString("\n")

    private fun signedHeaderNames(headers: Map<String, String>): String =
        headers.keys
            .map { it.lowercase(Locale.US) }
            .sorted()
            .joinToString(";")

    /**
     * S3 percent-encodes the object key once (unlike most services, which encode twice),
     * keeping `/` as the path separator.
     */
    internal fun encodeKey(key: String): String =
        key.split("/").joinToString("/") { segment ->
            val out = StringBuilder(segment.length)
            for (byte in segment.toByteArray(Charsets.UTF_8)) {
                val char = byte.toInt().toChar()
                if (byte >= 0 && UNRESERVED.indexOf(char) >= 0) {
                    out.append(char)
                } else {
                    out.append('%').append(String.format(Locale.US, "%02X", byte.toInt() and 0xFF))
                }
            }
            out.toString()
        }

    /** SigV4 trims the value and collapses runs of internal whitespace. */
    private fun normalizeHeaderValue(value: String): String =
        value.trim().replace(Regex("\\s+"), " ")

    private fun signingKey(
        secretAccessKey: String,
        dateStamp: String,
        region: String,
        service: String,
    ): ByteArray {
        val date = hmacSha256("AWS4$secretAccessKey".toByteArray(Charsets.UTF_8), dateStamp)
        val regional = hmacSha256(date, region)
        val serviceKey = hmacSha256(regional, service)
        return hmacSha256(serviceKey, TERMINATOR)
    }

    private fun hmacSha256(key: ByteArray, data: String): ByteArray {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        return mac.doFinal(data.toByteArray(Charsets.UTF_8))
    }

    fun sha256Hex(bytes: ByteArray): String = hex(MessageDigest.getInstance("SHA-256").digest(bytes))

    private fun hex(bytes: ByteArray): String {
        val out = StringBuilder(bytes.size * 2)
        for (byte in bytes) {
            out.append(String.format(Locale.US, "%02x", byte.toInt() and 0xFF))
        }
        return out.toString()
    }

    private fun amzDateFormat(): SimpleDateFormat =
        SimpleDateFormat("yyyyMMdd'T'HHmmss'Z'", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }

    private fun dateStampFormat(): SimpleDateFormat =
        SimpleDateFormat("yyyyMMdd", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }
}

File: ios/ClaraAssistant/Upload/AwsV4Signer.swift
import CryptoKit
import Foundation

/// Minimal AWS Signature Version 4 signer, scoped to exactly what the S3 uploader needs:
/// one request, a fully buffered payload, and no query string.
///
/// Hand-rolled rather than pulled from the AWS SDK for Swift, which would add a large package
/// graph (smithy-swift, aws-crt-swift) for a single PUT. Mirrors the Android implementation in
/// `com.metalens.app.upload.AwsV4Signer`; see AwsV4SignerTests for the AWS-published vector.
enum AwsV4Signer {
    private static let algorithm = "AWS4-HMAC-SHA256"
    private static let terminator = "aws4_request"

    /// RFC 3986 unreserved characters — everything else in a path segment is percent-encoded.
    private static let unreserved = Set(
        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~".utf8
    )

    /// Signs the request and returns the headers to add to it.
    ///
    /// `headers` are the extra headers to sign (content-type, x-amz-meta-*, x-amz-security-token).
    /// The returned dictionary also carries `x-amz-date`, `x-amz-content-sha256` and
    /// `Authorization`. `Host` is signed but not returned, because URLSession derives it from
    /// the URL itself.
    static func sign(
        method: String,
        host: String,
        /// Raw, *unencoded* object key with no leading slash.
        key: String,
        headers: [String: String],
        payloadSha256: String,
        accessKeyId: String,
        secretAccessKey: String,
        region: String,
        service: String = "s3",
        timestamp: Date
    ) -> [String: String] {
        let amzDate = amzDateFormatter.string(from: timestamp)
        let dateStamp = dateStampFormatter.string(from: timestamp)

        var signed = headers
        signed["host"] = host
        signed["x-amz-content-sha256"] = payloadSha256
        signed["x-amz-date"] = amzDate

        let canonical = canonicalRequest(
            method: method,
            key: key,
            headers: signed,
            payloadSha256: payloadSha256
        )
        let scope = "\(dateStamp)/\(region)/\(service)/\(terminator)"
        let toSign = stringToSign(amzDate: amzDate, scope: scope, canonicalRequest: canonical)

        let derivedKey = signingKey(
            secretAccessKey: secretAccessKey,
            dateStamp: dateStamp,
            region: region,
            service: service
        )
        let signature = hex(hmacSha256(key: derivedKey, data: Data(toSign.utf8)))

        let authorization = "\(algorithm) Credential=\(accessKeyId)/\(scope), "
            + "SignedHeaders=\(signedHeaderNames(signed)), "
            + "Signature=\(signature)"

        // Host is excluded: URLSession sets it from the URL, and setting it twice would break
        // the match.
        var result = signed.filter { $0.key.lowercased() != "host" }
        result["Authorization"] = authorization
        return result
    }

    static func canonicalRequest(
        method: String,
        key: String,
        headers: [String: String],
        payloadSha256: String
    ) -> String {
        let canonicalHeaders = headers
            .map { ($0.key.lowercased(), normalizeHeaderValue($0.value)) }
            .sorted { $0.0 < $1.0 }
            .map { "\($0.0):\($0.1)\n" }
            .joined()

        return [
            method,
            "/" + encodeKey(key),
            // No query string on a plain PUT Object.
            "",
            canonicalHeaders,
            signedHeaderNames(headers),
            payloadSha256,
        ].joined(separator: "\n")
    }

    static func stringToSign(amzDate: String, scope: String, canonicalRequest: String) -> String {
        [
            algorithm,
            amzDate,
            scope,
            sha256Hex(Data(canonicalRequest.utf8)),
        ].joined(separator: "\n")
    }

    private static func signedHeaderNames(_ headers: [String: String]) -> String {
        headers.keys
            .map { $0.lowercased() }
            .sorted()
            .joined(separator: ";")
    }

    /// S3 percent-encodes the object key once (unlike most services, which encode twice),
    /// keeping `/` as the path separator.
    static func encodeKey(_ key: String) -> String {
        key.split(separator: "/", omittingEmptySubsequences: false)
            .map { segment in
                var out = ""
                for byte in Array(segment.utf8) {
                    if unreserved.contains(byte) {
                        out.append(Character(UnicodeScalar(byte)))
                    } else {
                        out += String(format: "%%%02X", byte)
                    }
                }
                return out
            }
            .joined(separator: "/")
    }

    /// SigV4 trims the value and collapses runs of internal whitespace.
    private static func normalizeHeaderValue(_ value: String) -> String {
        value
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .split(whereSeparator: { $0.isWhitespace })
            .joined(separator: " ")
    }

    private static func signingKey(
        secretAccessKey: String,
        dateStamp: String,
        region: String,
        service: String
    ) -> Data {
        let date = hmacSha256(key: Data("AWS4\(secretAccessKey)".utf8), data: Data(dateStamp.utf8))
        let regional = hmacSha256(key: date, data: Data(region.utf8))
        let serviceKey = hmacSha256(key: regional, data: Data(service.utf8))
        return hmacSha256(key: serviceKey, data: Data(terminator.utf8))
    }

    private static func hmacSha256(key: Data, data: Data) -> Data {
        Data(HMAC<SHA256>.authenticationCode(for: data, using: SymmetricKey(data: key)))
    }

    static func sha256Hex(_ data: Data) -> String {
        hex(Data(SHA256.hash(data: data)))
    }

    private static func hex(_ data: Data) -> String {
        data.map { String(format: "%02x", $0) }.joined()
    }

    private static let amzDateFormatter: DateFormatter = utcFormatter("yyyyMMdd'T'HHmmss'Z'")
    private static let dateStampFormatter: DateFormatter = utcFormatter("yyyyMMdd")

    private static func utcFormatter(_ format: String) -> DateFormatter {
        let formatter = DateFormatter()
        formatter.dateFormat = format
        formatter.timeZone = TimeZone(identifier: "UTC")
        // Fixed locale: a user on a non-Gregorian calendar would otherwise produce dates AWS
        // cannot parse.
        formatter.locale = Locale(identifier: "en_US_POSIX")
        return formatter
    }
}
